torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -6,16 +6,134 @@ namespace rpc {
+      std::piecewise_construct,
+      std::forward_as_tuple(newTime),
+      std::forward_as_tuple(RpcRetryInfo(
+          to, std::move(message), fm, future, /* retryCount */ 1, maxRetries)));
+  rpcRetryMapCV_.notify_one();
+
+  return future;
+}
+
+void RpcAgent::retryExpiredRpcs() {
+  while (rpcAgentRunning_.load()) {
+    // TODO(osalpekar): This is temporary solution for handling empty map and
+    // sleeping until RPC retry duration is expired per Milestone 1.
+    std::unique_lock<std::mutex> lock(rpcRetryMutex_);
+    rpcRetryMapCV_.wait(lock, [&, this] {
+      return !(rpcAgentRunning_.load() && rpcRetryMap_.empty());
+    });
+
+    if (!rpcAgentRunning_.load()) {
+      return;
+    }
+    steady_clock_time_point earliestTimeout = rpcRetryMap_.begin()->first;
+    if (std::chrono::steady_clock::now() < earliestTimeout) {
+      std::this_thread::sleep_until(earliestTimeout);
+    }
+    RpcRetryInfo& earliestRpc = rpcRetryMap_.begin()->second;
+
+    // If the future for this RPC's last try was not completed, we will re-try
+    // sending the RPC and attach a callback to the returned future that marks
+    // the original future (which was returned to the user) as completed if the
+    // retry succeeds. We also compute a new time in the future (based on
+    // exponential backoff) at which this RPC would be re-tried next and add it
+    // to the rpcRetryMap_.
+    if (!earliestRpc.retryFuture_->completed()) {
+      auto fm = send(earliestRpc.to_, std::move(earliestRpc.message_));
+      fm->addCallback([earliestRpc](
+                          const rpc::Message& message,
+                          const c10::optional<utils::FutureError>& futErr) {
+        if (futErr) {
+          LOG(INFO) << "Send try "
+                    << std::to_string(earliestRpc.retryCount_ + 1)
+                    << " failed\n";
+        } else {
+          // This try succeeded, so we can mark the original future completed.
+          earliestRpc.originalFuture_->markCompleted(message);
+        }
+      });
+
+      // The exponential backoff algorithm being used here is:
+      // newTime = timeNow + (retryDuration * (backoffConstant ^ retryCount)).
+      std::chrono::milliseconds timedelta =
+          std::chrono::duration_cast<std::chrono::milliseconds>(
+              kRpcRetryDuration * pow(kBackoff, earliestRpc.retryCount_));
+      steady_clock_time_point newTime =
+          std::chrono::time_point_cast<std::chrono::milliseconds>(
+              std::chrono::steady_clock::now() + timedelta);
+
+      if (earliestRpc.retryCount_ + 1 < earliestRpc.maxRetries_) {
+        rpcRetryMap_.emplace(
+            std::piecewise_construct,
+            std::forward_as_tuple(newTime),
+            std::forward_as_tuple(RpcRetryInfo(
+                earliestRpc.to_,
+                std::move(earliestRpc.message_),
+                fm,
+                earliestRpc.originalFuture_,
+                earliestRpc.retryCount_ + 1,
```

Comment:
Similar to below comment - could we bump retryCount instead of constructing a new entry every time?